pkg/netpol/eval/exposure.go

```diff
@@ -0,0 +1,91 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package eval
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/eval/internal/k8s"
+	"github.com/np-guard/netpol-analyzer/pkg/netpol/internal/common"
+)
+
+// this file contains eval.PolicyEngine funcs which are related to exposure-analysis feature
+
+// addPotentiallyExposedPods adds pods which are potentially exposed to connect with (from/to) existing pods in the resources
+func (pe *PolicyEngine) addPotentiallyExposedPods() error {
+	// first adds a pod that represent connections with any namespace
+	_, err := pe.AddPodByNameAndNamespace(common.PodInExposedNs, common.AllNamespaces, nil)
+	if err != nil {
+		return err
+	}
+	// scan policies's rules : generate fake-exposure pod for rules with no match in the resources
+	for _, nsNetpolsMap := range pe.netpolsMap {
+		for pName, policy := range nsNetpolsMap {
+			// scan and handle policy rules which doesn't have a match in the resources
+			err = pe.addPodsForUnmatchedRules(pName, policy)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (pe *PolicyEngine) addPodsForUnmatchedRules(policyName string, policy *k8s.NetworkPolicy) error {
+	// gets the namespaceSelector-s from policy's rules which contain only namespaceSelector
+	namespacesOnly := policy.GetRulesSelectors()
+	// find if there are matches for the namespacesSelectors in pe.namespacesMap; if not add relevant pods
+	err := pe.addPodsForUnmatchedNamespaceSelectors(namespacesOnly, policyName)
+	// TODO : add pods for unmatched rules with:  only podSelector (in the policy's ns)
+	// - namespaceSelector + podSelector
+	return err
+}
+
+// gets a list of policy xgress rules consisted only from namespaceSelector.
+// adds new pod for each selector that does not have a matching namespace in the resources
+func (pe *PolicyEngine) addPodsForUnmatchedNamespaceSelectors(nsSelectors []*metav1.LabelSelector, policyName string) error {
+	for i, selector := range nsSelectors {
+		selectorMap, err := metav1.LabelSelectorAsMap(selector)
+		if err != nil {
+			return err
+		}
+		foundNs := pe.checkNamespaceSelectorsMatch(selectorMap)
+		if !foundNs {
+			_, err = pe.AddPodByNameAndNamespace(common.PodInExposedNs, common.NsNamePrefix+policyName+fmt.Sprint(i), selectorMap)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// getting a selector , checks if there is a namespace with labels containing all the selector's labels
+func (pe *PolicyEngine) checkNamespaceSelectorsMatch(reqSelector map[string]string) bool {
+	for _, ns := range pe.namspacesMap {
+		cnt := 0
+		// check if the selectors of the namespace contain the given selectors
+		for sel, val := range reqSelector {
+			if ns.Labels[sel] == val {
+				cnt++
+			}
+		}
+		if len(reqSelector) == cnt {
+			return true
+		}
+	}
```

Comment:
I think there are external functions to check if labels are matched by a selector, we do not need to implement this on our own.